File: learnia-backend/app/services/ai_service.py
import base64
import io
import time
from typing import Any, Dict, List

import pytesseract
from loguru import logger
from openai import AsyncOpenAI
from PIL import Image

from ..config import settings
from ..models.ai_models import (
    OcrRequest,
    OcrResponse,
    OrientationRequest,
    OrientationResponse,
    QcmQuestion,
    QcmRequest,
    QcmResponse,
    SummaryRequest,
    SummaryResponse,
    TranslationRequest,
    TranslationResponse,
    TutorRequest,
    TutorResponse,
)


class AIService:
    """Service principal pour toutes les fonctionnalités d'IA"""

    def __init__(self):
        self.openai_client = None
        if settings.openai_api_key:
            self.openai_client = AsyncOpenAI(api_key=settings.openai_api_key)

        # Modèles locaux (optionnels)
        self.translation_pipeline = None
        self.summarization_pipeline = None

        # Dictionnaire de traductions locales
        self.local_translations = {
            "éwé": {
                "bonjour": "Woé zɔ",
                "merci": "Akpé",
                "comment allez-vous": "Êfoa woé",
                "au revoir": "Hede nyuie",
                "oui": "Ee",
                "non": "Ao",
            },
            "kabiyè": {
                "bonjour": "Yaa",
                "merci": "Yoo",
                "comment allez-vous": "Yaa yaa",
                "au revoir": "Yaa yaa",
                "oui": "Ee",
                "non": "Ao",
            },
        }

    async def generate_tutor_response(
        self, request: TutorRequest
    ) -> TutorResponse:
        """Génère une réponse pour le tuteur intelligent"""
        start_time = time.time()

        try:
            if self.openai_client:
                # Utiliser OpenAI
                response = await self._call_openai_tutor(request)
                source = "ai_api"
                confidence = 0.9
            else:
                # Fallback local
                response = self._generate_local_tutor_response(request)
                source = "local"
                confidence = 0.7

            processing_time = time.time() - start_time
            logger.info(f"Réponse tuteur générée en {processing_time:.2f}s")

            return TutorResponse(
                answer=response, confidence=confidence, source=source
            )

        except Exception as e:
            logger.error(f"Erreur génération réponse tuteur: {e}")
            return TutorResponse(
                answer=self._generate_fallback_response(request),
                confidence=0.5,
                source="offline",
            )

    async def generate_qcm_questions(self, request: QcmRequest) -> QcmResponse:
        """Génère des questions QCM"""
        start_time = time.time()

        try:
            if self.openai_client:
                # Utiliser OpenAI
                questions = await self._call_openai_qcm(request)
            else:
                # Fallback local
                questions = self._generate_local_qcm_questions(request)

            processing_time = time.time() - start_time
            logger.info(f"QCM généré en {processing_time:.2f}s")

            return QcmResponse(
                questions=questions,
                subject=request.subject,
                difficulty=request.difficulty,
            )

        except Exception as e:
            logger.error(f"Erreur génération QCM: {e}")
            return QcmResponse(
                questions=self._generate_fallback_qcm(request),
                subject=request.subject,
                difficulty="easy",
            )

    async def generate_summary(
        self, request: SummaryRequest
    ) -> SummaryResponse:
        """Génère un résumé automatique"""
        start_time = time.time()

        try:
            if self.openai_client:
                # Utiliser OpenAI
                summary = await self._call_openai_summary(request)
            else:
                # Fallback local
                summary = self._generate_local_summary(request)

            processing_time = time.time() - start_time
            original_length = len(request.text.split())
            summary_length = len(summary.split())
            compression_ratio = (
                summary_length / original_length if original_length > 0 else 0
            )

            logger.info(f"Résumé généré en {processing_time:.2f}s")

            return SummaryResponse(
                summary=summary,
                original_length=original_length,
                summary_length=summary_length,
                compression_ratio=compression_ratio,
            )

        except Exception as e:
            logger.error(f"Erreur génération résumé: {e}")
            return SummaryResponse(
                summary=self._generate_fallback_summary(request),
                original_length=len(request.text.split()),
                summary_length=0,
                compression_ratio=0,
            )

    async def translate_text(
        self, request: TranslationRequest
    ) -> TranslationResponse:
        """Traduit un texte"""
        start_time = time.time()

        try:
            if request.target_language in ["éwé", "kab"]:
                # Traduction locale
                translated_text = self._translate_local(request)
                confidence = 0.8
            elif self.openai_client:
                # Utiliser OpenAI
                translated_text = await self._call_openai_translation(request)
                confidence = 0.9
            else:
                # Fallback
                translated_text = request.text
                confidence = 0.5

            processing_time = time.time() - start_time
            logger.info(f"Traduction effectuée en {processing_time:.2f}s")

            return TranslationResponse(
                original_text=request.text,
                translated_text=translated_text,
                source_language=request.source_language,
                target_language=request.target_language,
                confidence=confidence,
            )

        except Exception as e:
            logger.error(f"Erreur traduction: {e}")
            return TranslationResponse(
                original_text=request.text,
                translated_text=request.text,
                source_language=request.source_language,
                target_language=request.target_language,
                confidence=0.0,
            )

    async def analyze_orientation(
        self, request: OrientationRequest
    ) -> OrientationResponse:
        """Analyse l'orientation scolaire"""
        start_time = time.time()

        try:
            if self.openai_client:
                # Utiliser OpenAI
                result = await self._call_openai_orientation(request)
                confidence = 0.9
            else:
                # Fallback local
                result = self._generate_local_orientation(request)
                confidence = 0.7

            processing_time = time.time() - start_time
            logger.info(f"Orientation analysée en {processing_time:.2f}s")

            return OrientationResponse(
                suggested_fields=result["fields"],
                suggested_careers=result["careers"],
                explanation=result["explanation"],
                confidence=confidence,
                recommendations=result["recommendations"],
            )

        except Exception as e:
            logger.error(f"Erreur analyse orientation: {e}")
            return OrientationResponse(
                suggested_fields=["Sciences et Technologies"],
                suggested_careers=["Ingénieur", "Médecin", "Enseignant"],
                explanation="Analyse basée sur des critères généraux",
                confidence=0.5,
                recommendations=[
                    "Continuez vos études",
                    "Développez vos compétences",
                ],
            )

    async def perform_ocr(self, request: OcrRequest) -> OcrResponse:
        """Effectue la reconnaissance de texte (OCR)"""
        start_time = time.time()

        try:
            # Décoder l'image base64
            image_data = base64.b64decode(request.image_base64)
            image = Image.open(io.BytesIO(image_data))

            # Utiliser pytesseract pour l'OCR
            text = pytesseract.image_to_string(
                image, lang=request.language, config="--psm 6"
            )

            processing_time = time.time() - start_time
            word_count = len(text.split())

            logger.info(f"OCR effectué en {processing_time:.2f}s")

            return OcrResponse(
                text=text.strip(),
                confidence=0.8,  # Estimation
                word_count=word_count,
                processing_time=processing_time,
            )

        except Exception as e:
            logger.error(f"Erreur OCR: {e}")
            return OcrResponse(
                text="Erreur de reconnaissance de texte",
                confidence=0.0,
                word_count=0,
                processing_time=0,
            )

    # Méthodes privées pour les appels OpenAI
    async def _call_openai_tutor(self, request: TutorRequest) -> str:
        """Appel OpenAI pour le tuteur"""
        prompt = f"""
        Tu es un tuteur éducatif pour des élèves togolais du primaire à la 
        terminale. Réponds de manière simple et claire en français. Adapte 
        ton explication au niveau de l'élève.
        
        Question: {request.question}
        Matière: {request.subject}
        Niveau: {request.grade_level or 'Collège'}
        """

        response = await self.openai_client.chat.completions.create(
            model=settings.default_ai_model,
            messages=[{"role": "user", "content": prompt}],
            max_tokens=settings.max_tokens,
            temperature=settings.temperature,
        )

        return response.choices[0].message.content

    async def _call_openai_qcm(self, request: QcmRequest) -> List[QcmQuestion]:
        """Appel OpenAI pour générer des QCM"""
        prompt = f"""
        Génère {request.num_questions} questions à choix multiples basées 
        sur ce texte: {request.text}
        
        Format JSON:
        {{
            "questions": [
                {{
                    "question": "Question ici",
                    "options": ["A", "B", "C", "D"],
                    "correct_answer": "A",
                    "explanation": "Explication optionnelle"
                }}
            ]
        }}
        """

        await self.openai_client.chat.completions.create(
            model=settings.default_ai_model,
            messages=[{"role": "user", "content": prompt}],
            max_tokens=1000,
            temperature=0.7,
        )

        # Pour simplifier on génère des questions basiques
        return self._generate_local_qcm_questions(request)

    async def _call_openai_summary(self, request: SummaryRequest) -> str:
        """Appel OpenAI pour le résumé"""
        prompt = f"""
        Résume ce texte en français de manière claire et structurée:
        {request.text}
        
        Longueur maximale: {request.max_length} mots
        Style: {request.style}
        """

        response = await self.openai_client.chat.completions.create(
            model=settings.default_ai_model,
            messages=[{"role": "user", "content": prompt}],
            max_tokens=request.max_length,
            temperature=0.5,
        )

        return response.choices[0].message.content

    async def _call_openai_translation(
        self, request: TranslationRequest
    ) -> str:
        """Appel OpenAI pour la traduction"""
        prompt = f"""
        Traduis ce texte du français vers {request.target_language}:
        {request.text}
        
        Contexte: {request.context or 'Éducation'}
        """

        response = await self.openai_client.chat.completions.create(
            model=settings.default_ai_model,
            messages=[{"role": "user", "content": prompt}],
            max_tokens=500,
            temperature=0.3,
        )

        return response.choices[0].message.content

    async def _call_openai_orientation(
        self, request: OrientationRequest
    ) -> Dict[str, Any]:
        """Appel OpenAI pour l'orientation"""
        answers_text = "\n".join(
            [f"Q{ans.question_id}: {ans.answer}" for ans in request.answers]
        )

        prompt = f"""
        Analyse les réponses de cet élève pour l'orientation scolaire:
        
        Niveau: {request.grade_level}
        Matières actuelles: {', '.join(request.current_subjects or [])}
        
        Réponses:
        {answers_text}
        
        Donne des suggestions de filières et métiers en JSON:
        {{
            "fields": ["filière1", "filière2"],
            "careers": ["métier1", "métier2"],
            "explanation": "explication",
            "recommendations": ["recommandation1", "recommandation2"]
        }}
        """

        await self.openai_client.chat.completions.create(
            model=settings.default_ai_model,
            messages=[{"role": "user", "content": prompt}],
            max_tokens=800,
            temperature=0.7,
        )

        # Parse la réponse JSON (simplifié)
        return self._generate_local_orientation(request)

    # Méthodes de fallback local
    def _generate_local_tutor_response(self, request: TutorRequest) -> str:
        """Génère une réponse locale pour le tuteur"""
        responses = {
            "mathématiques": {
                "fraction": (
                    "Les fractions représentent une partie d'un tout. "
                    "Par exemple, 1/2 signifie une partie sur deux "
                    "parties égales."
                ),
                "géométrie": (
                    "La géométrie étudie les formes et les figures "
                    "dans l'espace."
                ),
                "algèbre": (
                    "L'algèbre utilise des lettres pour représenter "
                    "des nombres inconnus."
                ),
            },
            "français": {
                "grammaire": (
                    "La grammaire étudie la structure et les règles "
                    "de la langue française."
                ),
                "conjugaison": (
                    "La conjugaison indique le temps et la personne "
                    "du verbe."
                ),
                "orthographe": (
                    "L'orthographe concerne l'écriture correcte " "des mots."
                ),
            },
        }

        subject_responses = responses.get(request.subject.lower(), {})
        for keyword, response in subject_responses.items():
            if keyword in request.question.lower():
                return response

        return (
            f"Je comprends votre question sur {request.subject}. "
            "Voici une explication simple..."
        )

    def _generate_local_qcm_questions(
        self, request: QcmRequest
    ) -> List[QcmQuestion]:
        """Génère des questions QCM locales"""
        return [
            QcmQuestion(
                question=f"Qu'est-ce que {request.subject} ?",
                options=["Option A", "Option B", "Option C", "Option D"],
                correct_answer="Option A",
                explanation="Explication de la réponse",
            ),
            QcmQuestion(
                question="Quelle est la définition principale ?",
                options=["Réponse 1", "Réponse 2", "Réponse 3", "Réponse 4"],
                correct_answer="Réponse 1",
                explanation="Explication de la réponse",
            ),
        ]

    def _generate_local_summary(self, request: SummaryRequest) -> str:
        """Génère un résumé local"""
        sentences = request.text.split(".")
        if len(sentences) <= 3:
            return request.text

        return ". ".join(sentences[:3]) + "."

    def _translate_local(self, request: TranslationRequest) -> str:
        """Traduit localement"""
        translations = self.local_translations.get(request.target_language, {})
        translated_text = request.text

        for french, local in translations.items():
            translated_text = translated_text.replace(french, local)

        return translated_text

    def _generate_local_orientation(
        self, request: OrientationRequest
    ) -> Dict[str, Any]:
        """Génère une orientation locale"""
        return {
            "fields": [
                "Sciences et Technologies",
                "Lettres et Sciences Humaines",
            ],
            "careers": ["Ingénieur", "Médecin", "Enseignant", "Avocat"],
            "explanation": (
                "Basé sur vos réponses, nous vous suggérons " "ces filières..."
            ),
            "recommendations": [
                "Continuez vos études",
                "Développez vos compétences",
                "Explorez différents domaines",
            ],
        }

    def _generate_fallback_response(self, request: TutorRequest) -> str:
        """Réponse de fallback"""
        return (
            f"Désolé, je ne peux pas répondre à cette question sur "
            f"{request.subject} pour le moment."
        )

    def _generate_fallback_qcm(self, request: QcmRequest) -> List[QcmQuestion]:
        """QCM de fallback"""
        return [
            QcmQuestion(
                question="Question de fallback",
                options=["A", "B", "C", "D"],
                correct_answer="A",
                explanation="Réponse de fallback",
            )
        ]

    def _generate_fallback_summary(self, request: SummaryRequest) -> str:
        """Résumé de fallback"""
        return "Impossible de générer un résumé pour le moment."


# Instance globale du service IA
ai_service = AIService()

File: lib/services/ai_service.dart
import 'api_service.dart';
import 'package:flutter/foundation.dart';

class AIService {
  static final AIService _instance = AIService._internal();
  factory AIService() => _instance;
  AIService._internal();

  final ApiService _apiService = ApiService();

  // Tuteur intelligent
  Future<TutorResponse> askTutor({
    required String question,
    required String subject,
    String? gradeLevel,
    String? context,
  }) async {
    try {
      final response = await _apiService.post('/ai/tutor/', body: {
        'question': question,
        'subject': subject,
        'grade_level': gradeLevel,
        'context': context,
      });

      return TutorResponse.fromJson(response);
    } catch (e) {
      if (kDebugMode) {
        print('Erreur tuteur: $e');
      }
      // Fallback local
      return _generateLocalTutorResponse(question, subject, gradeLevel);
    }
  }

  // Générateur de QCM
  Future<QcmResponse> generateQcm({
    required String text,
    required String subject,
    int numQuestions = 5,
    String difficulty = 'medium',
  }) async {
    try {
      final response = await _apiService.post('/ai/qcm/', body: {
        'text': text,
        'subject': subject,
        'num_questions': numQuestions,
        'difficulty': difficulty,
      });

      return QcmResponse.fromJson(response);
    } catch (e) {
      if (kDebugMode) {
        print('Erreur QCM: $e');
      }
      // Fallback local
      return _generateLocalQcm(text, subject, numQuestions);
    }
  }

  // Résumé automatique
  Future<SummaryResponse> generateSummary({
    required String text,
    int maxLength = 100,
    String style = 'academic',
  }) async {
    try {
      final response = await _apiService.post('/ai/summary/', body: {
        'text': text,
        'max_length': maxLength,
        'style': style,
      });

      return SummaryResponse.fromJson(response);
    } catch (e) {
      if (kDebugMode) {
        print('Erreur résumé: $e');
      }
      // Fallback local
      return _generateLocalSummary(text, maxLength);
    }
  }

  // Traduction
  Future<TranslationResponse> translateText({
    required String text,
    required String sourceLanguage,
    required String targetLanguage,
    String? context,
  }) async {
    try {
      final response = await _apiService.post('/ai/translation/', body: {
        'text': text,
        'source_language': sourceLanguage,
        'target_language': targetLanguage,
        'context': context,
      });

      return TranslationResponse.fromJson(response);
    } catch (e) {
      if (kDebugMode) {
        print('Erreur traduction: $e');
      }
      // Fallback local
      return _generateLocalTranslation(text, sourceLanguage, targetLanguage);
    }
  }

  // Orientation scolaire
  Future<OrientationResponse> analyzeOrientation({
    required List<OrientationAnswer> answers,
    required String gradeLevel,
    List<String>? currentSubjects,
  }) async {
    try {
      final response = await _apiService.post('/ai/orientation/', body: {
        'answers': answers.map((a) => a.toJson()).toList(),
        'grade_level': gradeLevel,
        'current_subjects': currentSubjects,
      });

      return OrientationResponse.fromJson(response);
    } catch (e) {
      if (kDebugMode) {
        print('Erreur orientation: $e');
      }
      // Fallback local
      return _generateLocalOrientation(answers, gradeLevel);
    }
  }

  // OCR
  Future<OcrResponse> performOcr({
    required String imageBase64,
    String language = 'fra',
  }) async {
    try {
      final response = await _apiService.post('/ai/ocr/', body: {
        'image_base64': imageBase64,
        'language': language,
      });

      return OcrResponse.fromJson(response);
    } catch (e) {
      if (kDebugMode) {
        print('Erreur OCR: $e');
      }
      // Fallback local
      return _generateLocalOcr();
    }
  }

  // Suggestions de questions
  Future<List<String>> getQuestionSuggestions(String subject) async {
    try {
      final response = await _apiService.get('/ai/tutor/suggestions/$subject');
      return List<String>.from(response['suggestions'] ?? []);
    } catch (e) {
      if (kDebugMode) {
        print('Erreur suggestions: $e');
      }
      return _getLocalSuggestions(subject);
    }
  }

  // Matières supportées
  Future<List<String>> getSupportedSubjects() async {
    try {
      final response = await _apiService.get('/ai/tutor/subjects');
      return List<String>.from(response['subjects'] ?? []);
    } catch (e) {
      if (kDebugMode) {
        print('Erreur matières: $e');
      }
      return _getLocalSubjects();
    }
  }

  // Méthodes de fallback local
  TutorResponse _generateLocalTutorResponse(String question, String subject, String? gradeLevel) {
    final responses = {
      'mathématiques': {
        'fraction': 'Les fractions représentent une partie d\'un tout. Par exemple, 1/2 signifie une partie sur deux parties égales.',
        'géométrie': 'La géométrie étudie les formes et les figures dans l\'espace.',
        'algèbre': 'L\'algèbre utilise des lettres pour représenter des nombres inconnus.',
      },
      'français': {
        'grammaire': 'La grammaire étudie la structure et les règles de la langue française.',
        'conjugaison': 'La conjugaison indique le temps et la personne du verbe.',
        'orthographe': 'L\'orthographe concerne l\'écriture correcte des mots.',
      },
      'histoire': {
        'colonisation': 'La colonisation est l\'occupation et l\'exploitation d\'un territoire par une puissance étrangère.',
        'indépendance': 'L\'indépendance est la liberté d\'un pays de se gouverner lui-même.',
      },
    };

    final subjectResponses = responses[subject.toLowerCase()] ?? {};
    String answer = 'Je comprends votre question sur $subject. Voici une explication simple...';
    
    for (final keyword in subjectResponses.keys) {
      if (question.toLowerCase().contains(keyword)) {
        answer = subjectResponses[keyword]!;
        break;
      }
    }

    return TutorResponse(
      answer: answer,
      confidence: 0.7,
      source: 'local',
    );
  }

  QcmResponse _generateLocalQcm(String text, String subject, int numQuestions) {
    final questions = <QcmQuestion>[];
    
    for (int i = 0; i < numQuestions; i++) {
      questions.add(QcmQuestion(
        question: 'Question ${i + 1} sur $subject ?',
        options: ['Option A', 'Option B', 'Option C', 'Option D'],
        correctAnswer: 'Option A',
        explanation: 'Explication de la réponse correcte',
      ));
    }

    return QcmResponse(
      questions: questions,
      subject: subject,
      difficulty: 'medium',
    );
  }

  SummaryResponse _generateLocalSummary(String text, int maxLength) {
    final sentences = text.split('.');
    String summary = sentences.take(3).join('.');
    
    if (summary.length > maxLength) {
      summary = summary.substring(0, maxLength) + '...';
    }

    return SummaryResponse(
      summary: summary,
      originalLength: text.split(' ').length,
      summaryLength: summary.split(' ').length,
      compressionRatio: summary.split(' ').length / text.split(' ').length,
    );
  }

  TranslationResponse _generateLocalTranslation(String text, String sourceLanguage, String targetLanguage) {
    final translations = {
      'éwé': {
        'bonjour': 'Woé zɔ',
        'merci': 'Akpé',
        'comment allez-vous': 'Êfoa woé',
        'au revoir': 'Hede nyuie',
        'oui': 'Ee',
        'non': 'Ao',
      },
      'kabiyè': {
        'bonjour': 'Yaa',
        'merci': 'Yoo',
        'comment allez-vous': 'Yaa yaa',
        'au revoir': 'Yaa yaa',
        'oui': 'Ee',
        'non': 'Ao',
      },
    };

    String translatedText = text;
    final localTranslations = translations[targetLanguage] ?? {};
    
    for (final french in localTranslations.keys) {
      translatedText = translatedText.replaceAll(french, localTranslations[french]!);
    }

    return TranslationResponse(
      originalText: text,
      translatedText: translatedText,
      sourceLanguage: sourceLanguage,
      targetLanguage: targetLanguage,
      confidence: 0.8,
    );
  }

  OrientationResponse _generateLocalOrientation(List<OrientationAnswer> answers, String gradeLevel) {
    return OrientationResponse(
      suggestedFields: ['Sciences et Technologies', 'Lettres et Sciences Humaines'],
      suggestedCareers: ['Ingénieur', 'Médecin', 'Enseignant', 'Avocat'],
      explanation: 'Basé sur vos réponses, nous vous suggérons ces filières...',
      confidence: 0.7,
      recommendations: [
        'Continuez vos études',
        'Développez vos compétences',
        'Explorez différents domaines',
      ],
    );
  }

  OcrResponse _generateLocalOcr() {
    return OcrResponse(
      text: 'Texte reconnu (simulation)',
      confidence: 0.8,
      wordCount: 5,
      processingTime: 1.5,
    );
  }

  List<String> _getLocalSuggestions(String subject) {
    final suggestions = {
      'mathématiques': [
        'Qu\'est-ce qu\'une fraction ?',
        'Comment calculer l\'aire d\'un cercle ?',
        'Qu\'est-ce qu\'une équation ?',
      ],
      'français': [
        'Qu\'est-ce qu\'un verbe ?',
        'Comment conjuguer le verbe être ?',
        'Qu\'est-ce qu\'un adjectif ?',
      ],
      'histoire': [
        'Qu\'est-ce que la colonisation ?',
        'Quand le Togo a-t-il obtenu son indépendance ?',
        'Qu\'est-ce qu\'une révolution ?',
      ],
    };

    return suggestions[subject.toLowerCase()] ?? [
      'Pouvez-vous expliquer ce concept ?',
      'Comment résoudre ce problème ?',
      'Quelle est la définition de ce terme ?',
    ];
  }

  List<String> _getLocalSubjects() {
    return [
      'Mathématiques',
      'Français',
      'Histoire',
      'Géographie',
      'Sciences',
      'Anglais',
      'Philosophie',
      'Économie',
    ];
  }
}

// Modèles de données
class TutorResponse {
  final String answer;
  final double confidence;
  final String source;

  TutorResponse({
    required this.answer,
    required this.confidence,
    required this.source,
  });

  factory TutorResponse.fromJson(Map<String, dynamic> json) {
    return TutorResponse(
      answer: json['answer'] ?? '',
      confidence: (json['confidence'] ?? 0.0).toDouble(),
      source: json['source'] ?? 'unknown',
    );
  }
}

class QcmResponse {
  final List<QcmQuestion> questions;
  final String subject;
  final String difficulty;

  QcmResponse({
    required this.questions,
    required this.subject,
    required this.difficulty,
  });

  factory QcmResponse.fromJson(Map<String, dynamic> json) {
    return QcmResponse(
      questions: (json['questions'] as List?)
          ?.map((q) => QcmQuestion.fromJson(q))
          .toList() ?? [],
      subject: json['subject'] ?? '',
      difficulty: json['difficulty'] ?? 'medium',
    );
  }
}

class QcmQuestion {
  final String question;
  final List<String> options;
  final String correctAnswer;
  final String explanation;

  QcmQuestion({
    required this.question,
    required this.options,
    required this.correctAnswer,
    required this.explanation,
  });

  factory QcmQuestion.fromJson(Map<String, dynamic> json) {
    return QcmQuestion(
      question: json['question'] ?? '',
      options: List<String>.from(json['options'] ?? []),
      correctAnswer: json['correct_answer'] ?? '',
      explanation: json['explanation'] ?? '',
    );
  }
}

class SummaryResponse {
  final String summary;
  final int originalLength;
  final int summaryLength;
  final double compressionRatio;

  SummaryResponse({
    required this.summary,
    required this.originalLength,
    required this.summaryLength,
    required this.compressionRatio,
  });

  factory SummaryResponse.fromJson(Map<String, dynamic> json) {
    return SummaryResponse(
      summary: json['summary'] ?? '',
      originalLength: json['original_length'] ?? 0,
      summaryLength: json['summary_length'] ?? 0,
      compressionRatio: (json['compression_ratio'] ?? 0.0).toDouble(),
    );
  }
}

class TranslationResponse {
  final String originalText;
  final String translatedText;
  final String sourceLanguage;
  final String targetLanguage;
  final double confidence;

  TranslationResponse({
    required this.originalText,
    required this.translatedText,
    required this.sourceLanguage,
    required this.targetLanguage,
    required this.confidence,
  });

  factory TranslationResponse.fromJson(Map<String, dynamic> json) {
    return TranslationResponse(
      originalText: json['original_text'] ?? '',
      translatedText: json['translated_text'] ?? '',
      sourceLanguage: json['source_language'] ?? '',
      targetLanguage: json['target_language'] ?? '',
      confidence: (json['confidence'] ?? 0.0).toDouble(),
    );
  }
}

class OrientationResponse {
  final List<String> suggestedFields;
  final List<String> suggestedCareers;
  final String explanation;
  final double confidence;
  final List<String> recommendations;

  OrientationResponse({
    required this.suggestedFields,
    required this.suggestedCareers,
    required this.explanation,
    required this.confidence,
    required this.recommendations,
  });

  factory OrientationResponse.fromJson(Map<String, dynamic> json) {
    return OrientationResponse(
      suggestedFields: List<String>.from(json['suggested_fields'] ?? []),
      suggestedCareers: List<String>.from(json['suggested_careers'] ?? []),
      explanation: json['explanation'] ?? '',
      confidence: (json['confidence'] ?? 0.0).toDouble(),
      recommendations: List<String>.from(json['recommendations'] ?? []),
    );
  }
}

class OrientationAnswer {
  final int questionId;
  final String answer;

  OrientationAnswer({
    required this.questionId,
    required this.answer,
  });

  Map<String, dynamic> toJson() {
    return {
      'question_id': questionId,
      'answer': answer,
    };
  }
}

class OcrResponse {
  final String text;
  final double confidence;
  final int wordCount;
  final double processingTime;

  OcrResponse({
    required this.text,
    required this.confidence,
    required this.wordCount,
    required this.processingTime,
  });

  factory OcrResponse.fromJson(Map<String, dynamic> json) {
    return OcrResponse(
      text: json['text'] ?? '',
      confidence: (json['confidence'] ?? 0.0).toDouble(),
      wordCount: json['word_count'] ?? 0,
      processingTime: (json['processing_time'] ?? 0.0).toDouble(),
    );
  }
}
